Refactor useListenForPostMessage

cellog opened the pull request:
# Description

This refactors `useListenForPostMessage` to do several new things:

1. restrict `event.data` values to be similar to a redux action. They are an object with indices `type` and `payload`.
2. the hook accepts as parameters the type, an optional validator for the payload, and an optional default value
3. it can now simply be used in a single line as in:

```js
const scrollPosition = useListenForPostMessage(window, {
  type: POST_MESSAGE_SCROLL_POSITION,
  validator: isPositiveInteger,
  default: 0
})
```

The value will trigger a re-render only if the received message is secure, and matches the expected message format, AND the value has actually changed. This will result in less jank for the scrolling stuff (probably, can't test because no iphone).

In addition, this test completely refactors the test for `useListenForPostMessage` to follow best practices. Instead of using `testHook`, the test creates a simple functional component that actually uses the hook, and uses its output to populate a text field that it asserts against. It also takes advantage of Jest's new ability to run a test against multiple inputs (https://github.com/sapegin/jest-cheat-sheet#data-driven-tests-jest-23) to greatly simplify the actual test file.

A small update to the paywall is necessary to move to the new format for scrollPosition, and this PR also takes advantage of the chance to make the postMessage more secure and only send it to the paywall's iframe.

# Issues

<!-- This PR should fix or reference at least one existing issue ID. Add or delete as appropriate. -->
Fixes #1816 

# Checklist:

- [X] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [X] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [X] My code follows the style guidelines of this project, including naming conventions
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [X] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

<!--
PS: [Read how to write the perfect pull request](https://blog.github.com/2015-01-21-how-to-write-the-perfect-pull-request/)
-->

cnasc commented on paywall/src/hooks/browser/useListenForPostMessage.js:
```diff
@@ -2,21 +2,32 @@ import { useEffect, useState } from 'react'
+    if (typeof event.data.type !== 'string') return
+    if (event.data.type !== type) return
+    // optional validator
+    if (!validator || (validator && validator(event.data.payload))) {
```
`validator` is a function that returns truthy for data we want? Do you have an example of how that would be used in practice? Just curious, it seems like a good thing to have but I can't come up with anything concrete at the moment.

LG!

cellog replied:
yep. The most immediate use is ensuring that scrollPosition is always a positive integer, so the existing `unlock-app/utils/validators.js` function `isPositiveInteger` can be used as a validator in that instance. That's why I chose this specific example in the description:

```js
const scrollPosition = useListenForPostMessage(window, {
  type: POST_MESSAGE_SCROLL_POSITION,
  validator: isPositiveInteger, // this function only validates numbers >= 0
  default: 0
})
```

cnasc replied:
D'oh, can't believe I skimmed right over that. Makes sense!

julien51 commented:
Ha! Do you mind using a better PR name please ;)

julien51 commented on paywall/src/__tests__/hooks/browser/useListenForPostMessage.test.js:
```diff
@@ -28,164 +32,145 @@ describe('useListenForPostMessage hook', () => {
-    rtl.testHook(() => useListenForPostMessage(fakeWindow), {
-      wrapper,
+  function MockListener({
+    /* eslint-disable react/prop-types */
```
Any reason you need to disable these?

cellog replied:
Just seems weird to define prop types for a test helper, but I can do this if you'd like

cellog commented:
> Ha! Do you mind using a better PR name please ;)

Whatever do you mean? ;)

julien51 commented on paywall/src/__tests__/hooks/browser/useListenForPostMessage.test.js:
```diff
@@ -28,164 +32,145 @@ describe('useListenForPostMessage hook', () => {
-  it('ignores calls on the server', () => {
-    expect.assertions(1)
+  describe('does nothing in invalid contexts', () => {
+    it.each([
```
👍

cellog replied:
yeah isn't this cool? could definitely slim up some of the tests like CreatorLockForm validation

julien51 commented on paywall/src/hooks/browser/useListenForPostMessage.js:
```diff
@@ -2,21 +2,32 @@ import { useEffect, useState } from 'react'
 import { getRouteFromWindow } from '../../utils/routes'
 
-export default function useListenForPostmessage(window) {
+export default function useListenForPostmessage(
```
Do you mind adding JS doc comments to tell us a bit more about that this hook does?

cellog replied:
done

julien51 replied:
wow, great job!

julien51 replied:
Nit: do you mind using camelCase for this? (I see the previous code was indeed using the wrong case, but let's fix it here! Thanks!

cellog commented:
OK, this is ready for a 2nd look. I discovered a bug in the passing of origin to the iframe `postMessage` call, in that it was passing the full paywall URL, which is `https://unlock-protocol.com/paywall` currently. Postmessage expects an origin, which would be `https://unlock-protocol.com`, so I added in a call to `new URL(paywallUrl).origin` in order to retrieve the origin.

While working to test this change, I discovered that there was a gap in coverage for `scrollLoop` inside `paywall-builder/build.js`, and so I pulled it out of the `buildPaywall` function to make this task easier, and added 6 new tests to cover the use cases. I considered splitting this out, but it is a pretty minor change, and would require doing a pre-PR with a buggy, insecure buildPaywall, and then immediately this one would change almost as many lines to fix it.

In any case, PTAL on Monday (or earlier if you're revved up)

julien51 commented on paywall/src/hooks/browser/useListenForPostMessage.js:
```diff
@@ -2,21 +2,72 @@ import { useEffect, useState } from 'react'
+    // origin is passed in by the paywall, see paywall-builder/build.js
     const { origin } = getRouteFromWindow(window)
+    // **SECURITY CHECKS**
+
```
nit: remove line?
